Improve pygobject support

virtuald opened the pull request:
* Add Gtk, GDK, etc
* Bring in library dependencies automatically
* Detect overrides automatically

One thing that would be *really* nice if there were a way to say "anytime gi.repository.* is imported, call some hook" which could use the `get_gi_typelibs()` function to (usually) automatically support that library.

I haven't tested this on Windows yet, but I'll be doing that in a few minutes.

virtuald commented:
Doesn't work on Windows.

virtuald commented:
Ok, I tested it on Windows (well.. wine to be precise) and it works great. It also works on Linux.

One problem left is that for GTK it doesn't copy over things like themes, icons, etc, so it won't actually work until someone copies that stuff over.

virtuald commented:
Ok, so I looked at this for a little while, and I'm not going to bother fixing the Gtk theme problem. It's easy enough to copy the directories manually -- and it's a pain to figure out what exactly you need to copy and override it. 

Instead, I added a warning about it. This should be ready for review/merge, as it fixes most of the problem and improves pygobject support in pyinstaller.

htgoebel commented on PyInstaller/depend/bindepend.py:
```diff
@@ -880,3 +880,17 @@ def get_python_library_path():
+    if is_unix:
+        return findLibrary(name)
+    elif is_win:
+        return getfullnameof(name)
```
This "loads" dependencies? I assume this should be "searches for".

matysek replied:
Look at func `mac_set_relative_dylib_deps(libname, distname)` in `dylib.py`. It could have some inspiration.

virtuald replied:
Cool, I'll do that. Haven't gotten around to my OSX port of my app, but hopefully soon.

htgoebel commented on PyInstaller/hooks/hook-gi.repository.GIRepository.py:
```diff
@@ -0,0 +1,15 @@
+
+from PyInstaller.utils.hooks import get_gi_typelibs
+
+binaries, datas, hiddenimports = get_gi_typelibs('GIRepository', '2.0')
```
Please add line-feed to the end of the last line.

htgoebel commented on PyInstaller/hooks/hook-gi.repository.GdkPixbuf.py:
```diff
@@ -0,0 +1,15 @@
+
+from PyInstaller.utils.hooks import get_gi_typelibs
+
+binaries, datas, hiddenimports = get_gi_typelibs('GdkPixbuf', '2.0')
```
Please add line-feed to the end of the last line.

htgoebel commented on PyInstaller/hooks/hook-gi.repository.Pango.py:
```diff
@@ -0,0 +1,15 @@
+
+from PyInstaller.utils.hooks import get_gi_typelibs
+
+binaries, datas, hiddenimports = get_gi_typelibs('Pango', '1.0')
```
Please add line-feed to the end of the last line.

htgoebel commented on PyInstaller/utils/hooks/__init__.py:
```diff
@@ -1015,26 +1017,69 @@ def relpath_to_config_or_make(filename):
 
 
 def get_typelibs(module, version):
+    '''deprecated; only here for backwards compat'''
```
Please log a deprecation warning here.

htgoebel commented on PyInstaller/utils/hooks/__init__.py:
```diff
@@ -1015,26 +1017,69 @@ def relpath_to_config_or_make(filename):
-    typelibs_path = eval_statement(statement)
-    if not typelibs_path:
+repo = GIRepository.Repository.get_default()
+module, version = ('%s', '%s')
```
Should be `%r` (without quotes), do ensure correct quoting. Otherwise a string like `funny'string` will break (much like SQL-injection :-)

htgoebel commented:
Thanks for this pull-request. It's a good improvement, since it eases developing new GI hooks. I added some line-comments, though.

htgoebel commented on PyInstaller/utils/hooks/__init__.py:
```diff
@@ -1015,26 +1017,69 @@ def relpath_to_config_or_make(filename):
+def get_gi_typelibs(module, version):
+    """
+    Returns a tuple of (binaries, datas, hiddenimports) to be used by PyGObject
+    related hooks. Loads dependencies recursively also.
```
This "loads" dependencies? I assume this should be "searches for".

htgoebel commented on PyInstaller/utils/hooks/__init__.py:
```diff
@@ -1015,26 +1017,69 @@ def relpath_to_config_or_make(filename):
+    return binaries, datas, hiddenimports
+    
+def _get_gi_typelibs(module, version, datas, binaries, hiddenimports):
+    
```
Please add a bit of documentation here. Thanks.

htgoebel commented on PyInstaller/utils/hooks/__init__.py:
```diff
@@ -827,13 +828,14 @@ def get_package_paths(package):
 
     :param pattern: String pattern to match only submodules containing
                     this pattern in the name.
+    :param exact: If True, will only make exact match on submodule names
```
Does not match the parameter-list.

virtuald commented:
I fixed the line feed issues and rebased + pushed, but github isn't showing your comments as outdated.

The remaining issue to resolve is adjusting findSystemLibrary() for darwin. If you have ideas on how to address that, that would be awesome. Otherwise, I'll play with it some tonight.

virtuald commented:
I've been working on the OSX piece this weekend, and PyGObject 3.18 depends on https://github.com/pyinstaller/pyinstaller/pull/1689 because of its use of `pkgutil.get_loader()`.

virtuald commented on PyInstaller/depend/bindepend.py:
```diff
@@ -873,3 +873,19 @@ def get_python_library_path():
+    elif is_win:
+        return getfullnameof(name)
+    else:
+        raise ValueError("findSystemLibrary not implemented for this platform")
```
Regarding the OSX version of this, I'm not quite sure how to solve it. However, when I call this on OSX, GIRepository is giving me a full absolute path, so it never hits this codepath anyways.

htgoebel replied:
Okay, so we leave it this way for now. Thanks for taking the time analyzing this.

virtuald commented:
This current update has been tested on OSX and works with a fairly complex application (as long as #1689 is applied also!). I need to retest this with windows, but I don't anticipate any issues.
